src/coreclr/src/gc/gc.cpp

```diff
@@ -2084,46 +2084,47 @@ uint8_t* tree_search (uint8_t* tree, uint8_t* old_address);
 #ifdef USE_INTROSORT
 #define _sort introsort::sort
 #elif defined(USE_VXSORT)
-#define _sort vxsort
-namespace std
+#define _sort do_vxsort
+#ifdef USE_VXSORT
+void do_vxsort(uint8_t** low, uint8_t** high, unsigned int depth)
 {
-    template <class _Ty>
-    class numeric_limits
+    assert(GCToEEInterface::HasInstructionSet(kInstructionSetAVX2));
+    if (GCToEEInterface::HasInstructionSet(kInstructionSetAVX512))
     {
-    public:
-        static _Ty Max()
-        {
-            return _Ty();
-        }
-    };
-
-    template <>
-    class numeric_limits<int64_t>
+        do_vxsort_avx512(low, high);
```

Comment:
Are there any numbers showing how AVX512 performs in real world scenarios?

It is fairly well documented that using AVX512 *can* cause downclocking which in certain circumstances can lead to decreased overall application performance (AVX2 can as well, but to a much lesser extent).
* The impact is lower on newer processors such as `Ice Lake` as compared to older `Skylake Server`
* The impact is typically relegated to instructions that use the floating-point or multiplication (including FMA) pipelines
* The impact is typically for 256 and 512-bit variants of the instructions, rather than 128-bit

The official guidance from Intel is the following (from the May 2020 edition of [Intel 64 and IA-32 Architectures Optimization Reference Manual](https://software.intel.com/content/www/us/en/develop/download/intel-64-and-ia-32-architectures-optimization-reference-manual.html)):
![image](https://user-images.githubusercontent.com/10487869/85576402-3b994b00-b5ed-11ea-81ab-5b02ac5e5ac1.png)

The section detailing `Skylake Server Power Management` is a bit longer and includes more details on what impacts each frequency level and includes various graphs/examples of the typical impact, with the explicit guidance to still profile.

Reply:
Downclocking can be pretty severe on 14nm parts right now:
From Anandtech's article [Sizing Up Servers: Intel's Skylake-SP Xeon versus AMD's EPYC 7000 - The Server CPU Battle of the Decade?](https://www.anandtech.com/show/11544/intel-skylake-ep-vs-amd-epyc-7000-cpu-battle-of-the-decade/8)
![image](https://user-images.githubusercontent.com/125730/85596303-4639fb80-b652-11ea-8eb7-b7cf96683dba.png)

While that is true, we are still seeing substantial perf boost on a given machine (copy pasting some preliminary results, while I'm still improving the code behind the scenes):

For 64-bit sorting, here is the original + 8way unrolled results, obtained from a shared instance running a Xeon Silver 4216 with nominal speed @ 2.1Ghz:

```
---------------------------------------------------------------------------------------------------
Benchmark (<type/vector-isa/unroll>/size/threads)                      Time/N (per Element)
---------------------------------------------------------------------------------------------------
BM_full_introsort/65536/threads:1                                      56.55590 ns
BM_full_introsort/131072/threads:1                                     60.63450 ns
BM_full_introsort/262144/threads:1                                     65.21260 ns
BM_full_introsort/524288/threads:1                                     69.03150 ns
BM_full_introsort/1048576/threads:1                                    73.16970 ns

BM_vxsort<int64_t, vector_machine::AVX2, 8>/65536/threads:1            20.14540 ns
BM_vxsort<int64_t, vector_machine::AVX2, 8>/131072/threads:1           21.08380 ns
BM_vxsort<int64_t, vector_machine::AVX2, 8>/262144/threads:1           22.36290 ns
BM_vxsort<int64_t, vector_machine::AVX2, 8>/524288/threads:1           23.05820 ns
BM_vxsort<int64_t, vector_machine::AVX2, 8>/1048576/threads:1          24.21020 ns

BM_vxsort<int64_t, vector_machine::AVX512, 8>/131072/threads:1          8.23500 ns
BM_vxsort<int64_t, vector_machine::AVX512, 8>/262144/threads:1          8.90877 ns
BM_vxsort<int64_t, vector_machine::AVX512, 8>/524288/threads:1          9.94376 ns
BM_vxsort<int64_t, vector_machine::AVX512, 8>/1048576/threads:1        10.70730 ns
```

So it's pretty clear the AVX512 is out-performing AVX2, post-downclocking.
There is also a very clear reason for this staggering improvement, if you are aware of two key points:
* AVX2 is missing some int64 functionality (certain int64 ops are more expensive)
* There is an extra specific AVX512 intrinsic (_mm512_compress_storeu_epiXX), which is removing the entire lookup table + cache reference that was involved for loading the lookup entry (which is a huge win)

As such, the perf bump can be thought of as *ONLY* being 2x-ish due to downclocking, with a clear expectation of seeing 3x with future 10nm parts and below.

As for int32 (which is also probably going to be part of this PR, given that we will probably dynamically switch to using int32 for smaller mark-lists):

```
---------------------------------------------------------------------------------------------------
Benchmark (<type/vector-isa/unroll>/size/threads)                      Time/N (per Element)
---------------------------------------------------------------------------------------------------
BM_vxsort<int32_t, vector_machine::AVX2, 8>/65536/threads:1            5.47985 ns
BM_vxsort<int32_t, vector_machine::AVX2, 8>/131072/threads:1           5.89806 ns
BM_vxsort<int32_t, vector_machine::AVX2, 8>/262144/threads:1           6.13415 ns
BM_vxsort<int32_t, vector_machine::AVX2, 8>/524288/threads:1           6.49738 ns
BM_vxsort<int32_t, vector_machine::AVX2, 8>/1048576/threads:1          7.10343 ns
 
BM_vxsort<int32_t, vector_machine::AVX512, 8>/65536/threads:1          3.70457 ns
BM_vxsort<int32_t, vector_machine::AVX512, 8>/131072/threads:1         3.94963 ns
BM_vxsort<int32_t, vector_machine::AVX512, 8>/262144/threads:1         4.12517 ns
BM_vxsort<int32_t, vector_machine::AVX512, 8>/524288/threads:1         4.35658 ns
BM_vxsort<int32_t, vector_machine::AVX512, 8>/1048576/threads:1        4.86922 ns
```

While the perf boost from AVX2 to AVX512 is smaller for int32, it is still substantial. Also, please remember these are preliminary results that are somewhat less than optimal given the lack of direct HW I'm experiencing, and that fact that everything I do with AVX512F is simply more painful because of this.

The benchmarks and code are all up-to-date and accessible from the [vxsort-cpp](https://github.com/damageboy/vxsort-cpp)

Finally, there is also a discussion to be had about how long do the downclocking effects linger on **POST** sorting.
As to that, the long answer is that Travis Downs (@trav_downs) measured this independently: https://travisdowns.github.io/blog/2020/01/17/avxfreq1.html

The short answer is no more than ~700usec:

What | Time | Description
-- | -- | -- | 
Voltage Transition | ~8 to 20 μs | Time required for a voltage transition, depends on the frequency 
Frequency Transition | ~10 μs | Time required for the halted part of a frequency transition 
Relaxation Period | ~680 μs | Time required to go back to a lower power license, measured from the last instruction requiring the higher license 

It is important to note that on server CPUs, this normally applies for individual cores, where on client CPUs this is "global".
For 14nm there are currently only server parts, with the exception of Intel Icelake CPUs.

I think that given we are saving multiple milliseconds with this optimization, even while increasing the mark-list, and paying with 700usec of reduced frequency post that time-saving window directly is more than a reasonable and temporary sacrifice.

My conclusion from this is that the current "state of the union" is roughly the worst it can get (although multi-threaded results are missing, due to me not having access to a dedicated machine where I could produce meaningful results: I barely respect the results I have, in that sense since this is not a machine I can kick people off of).

As you mentioned, Icelake and Tigerlake make things only better, and it is hard to imagine AMD doing worse once they come around to supporting AVX512, probably sometime around transitioning to TSMC 5nm.

I still think I can kick things up with a few more iterations (e.g. coming weekends). Peter has inspired me with an idea that can truly go ballistic, and I have some more ammo left regardless. So this really is as "bad" as it gets.

Hope this helps clear the fog and justify the risk.

Reply:
The main thing I'm worried about is "micro-benchmark" vs "real-world".

Its pretty easy to define a microbenchmark that shows AVX-512 or AVX2 is faster. The concern is then how that impacts everything that runs after that is done.

For example, given the numbers above AVX-512 is 2-5x faster (depending on if compared to AVX2 or SSE). However, that is ~10ns vs ~50ns. If the downclocking can last up to 700 microseconds, that is 14000x longer than the SSE scenario: https://www.google.com/search?q=700+microseconds+%2F+50+nanoseconds and the question is whether the gains outweigh potentially running the entire processor at 50-75% speed for that much longer.
* This is naturally assuming a worst or near worst case scenario, it may not actually be that bad; I just think it is something that we should carefully measure and with real world workloads, rather than just micro-benchmarks

Reply:
Well, it's 10ns vs 70ns **per element**.
Multiplied by 2<sup>20</sup> it is... more...

Agreed that the only thing that counts in the end of the day is real life perf for the end-to-end case rather than benchmarks. But the more the benchmarks end up being overwhelmingly extreme compared to the starting point, at some point even real life's ability to mess with you comes to an end.

Sorting, and in these sizes (more importantly) tends to always go to the extreme side of the spectrum for the "Optimize at all costs" arguments.

We can hopefully allow larger and larger mark-lists with this, which as far as I can tell has all kinds of all-around healing effects, though I'm not a GC guru.

Reply:
There is a real world motivation here - we have seen real world scenarios where mark list sorting takes several milliseconds (with ~100k mark list entries). We currently have a limitation at 150k entries - when we run out, we have to take a less efficient path in plan_phase, which runs into the tens of milliseconds for real world scenarios that I've seen.

I will consider limiting the AVX512 sorting to large mark lists though, to make sure we get a sizeable benefit.